tests/src/test/scala/com/nvidia/spark/rapids/ParquetWriterSuite.scala

```diff
@@ -118,4 +133,111 @@ class ParquetWriterSuite extends SparkQueryCompareTestSuite {
+    when(mockDtype.getSizeInBytes).thenReturn(1024)
+    val mockDataType = mock(classOf[DataType])
+    val spyGpuCol = spy(GpuColumnVector.from(spyCol, ByteType))
+    when(spyCol.getType()).thenReturn(mockDtype)
+    when(spyGpuCol.dataType()).thenReturn(mockDataType)
+    when(mockDataType.defaultSize).thenReturn(1024)
+
+    (spyCol, spyGpuCol)
+  }
+
+  val _2GB = 2L * 1024 * 1024 * 1024
+  val APPROX_PAR_META_DATA = 10 * 1024 * 1024 // we are estimating 10MB
+  val BYTES_ALLOWED_PER_BATCH = _2GB - APPROX_PAR_META_DATA
+
+  private def whenSplitCalled(cb: ColumnarBatch): Unit = {
+    val rows = cb.numRows()
+    val eachRowSize = cb.numCols() * 1024
+    val rowsAllowedInABatch = BYTES_ALLOWED_PER_BATCH / eachRowSize
+    val spillOver = cb.numRows() % rowsAllowedInABatch
+    val splitRange = scala.Range(rowsAllowedInABatch.toInt, rows, rowsAllowedInABatch.toInt)
+    scala.Range(0, cb.numCols()).indices.foreach { i =>
+      val spyCol = cb.column(i).asInstanceOf[GpuColumnVector].getBase
+      val splitCols0 = scala.Range(0, splitRange.length).map { _ =>
+        val spySplitCol = spy(ColumnVector.fromBytes(4, 5, 6))
+        when(spySplitCol.getRowCount()).thenReturn(rowsAllowedInABatch)
+        spySplitCol
+      }
+      val splitCols = if (spillOver > 0) {
+        val splitCol = spy(ColumnVector.fromBytes(3))
+        when(splitCol.getRowCount()).thenReturn(spillOver)
+        splitCols0 :+ splitCol
+      } else {
+        splitCols0
+      }
+      when(spyCol.split(any())).thenReturn(splitCols.toArray)
+    }
+  }
+
+  private var compressWithParquetMethod: Option[Method] = None
+  private var parquetSerializerInstance: Option[Any] = None
+
+  private def testCompressColBatch(
+     cudfCols: Array[ColumnVector],
+     gpuCols: Array[org.apache.spark.sql.vectorized.ColumnVector]): Unit = {
+    if (!withCpuSparkSession(s => s.version < "3.1.0")) {
+
+      // mock static method for Table
+      val theTableMock = mockStatic(classOf[Table], (_: InvocationOnMock) =>
+        new TableWriter {
+          override def write(table: Table): Unit = {
+            val tableSize = table.getColumn(0).getType.getSizeInBytes * table.getRowCount
+            if (tableSize > Int.MaxValue) {
+              fail(s"Parquet file went over the allowed limit of $BYTES_ALLOWED_PER_BATCH")
+            }
+          }
+
+          override def close(): Unit = {
+            // noop
+          }
+        })
+
+      withResource(cudfCols) { _=>
+        val cb = new ColumnarBatch(gpuCols, ROWS)
+        whenSplitCalled(cb)
+        try {
+          val method = compressWithParquetMethod.getOrElse {
+            val classOfSerializer = Class.forName(
+              "com.nvidia.spark.rapids.shims.spark310.ParquetCachedBatchSerializer")
+            parquetSerializerInstance = Some(classOfSerializer.newInstance())
+            val compressWithParquet =
+              classOfSerializer.getMethod("compressColumnarBatchWithParquet",
```

Comment:
Why is this using reflection to call code we wrote?

Reply:
Because the code won't compile against Spark 3.0 if we directly reference the `ParquetCachedBatchSerializer`. 

This is why we need a test framework with which we can ignore tests at compile time.

Reply:
ah right the tests problem.  that too will be a followup.